lib/data/in_memory/backend.js

```diff
@@ -10,14 +9,26 @@ export function resetCount() {
 }
 
 export const backend = {
-    put: function putMem(request, keyContext, reqUids, callback) {
-        let value = new Buffer(0);
-        request.on('data', data => value = Buffer.concat([value, data]))
+    put: function putMem(request, size, keyContext, reqUids, callback) {
+        const value = new Buffer(size);
+        let cursor = 0;
+        let exceeded = false;
+        request.on('data', data => {
+            if (cursor + data.length > size) {
+                exceeded = true;
+            }
+            if (!exceeded) {
+                data.copy(value, cursor);
+            }
+            cursor += data.length;
+        })
         .on('end', () => {
-            request.calculatedHash = crypto.createHash('md5')
-                                           .update(value).digest('hex');
-            ds[count] = { value, keyContext };
-            callback(null, count++);
+            if (exceeded) {
+                callback('Data stream exceed announced size');
```

Comment:
We use arsenal errors as the error in callbacks, strings are not meaningful enough.